routing/payment_lifecycle.go

```diff
@@ -128,34 +259,37 @@ lifecycle:
 			return [32]byte{}, nil, err
 		}
 
-		// We start every iteration by fetching the lastest state of
-		// the payment from the ControlTower. This ensures that we will
-		// act on the latest available information, whether we are
-		// resuming an existing payment or just sent a new attempt.
-		payment, err := p.router.cfg.Control.FetchPayment(
-			p.identifier,
-		)
-		if err != nil {
-			return [32]byte{}, nil, err
-		}
-
-		// Using this latest state of the payment, calculate
-		// information about our active shards and terminal conditions.
-		state, err := p.paymentState(payment)
+		// We update the payment state on every iteration. Since the
+		// payment state is affected by multiple goroutines (ie,
+		// collectResultAsync), it is NOT guaranteed that we always
+		// have the latest state here. This is fine as long as the
+		// state is consistent as a whole.
+		//
+		// Acquire a lock and update the payment state, then make a
+		// copy of the state. We will act upon the copied state for the
+		// rest of the loop.
+		p.state.Lock()
```

Comment:
Hm, I'm don't really see why this would help. The original idea here is that the state is "consistent" the moment you fetch the payment from the database. That some other goroutine alters it before we have  acted on it shouldn't really matter.

Even after this change there could still be a go routine altering the state after we release the lock, and then we are back to acting on an "outdated" state.

Reply:
>  In particular, the paymentState.terminate might become inconsistent with the rest

Can you explain the scenario where it would become inconsistent?

Reply:
Commented below with a scenario. I think it's ok to act on an "outdated" state, just not an inconsistent one. These two are different as, if I see a shard fails, I need to know whether I want to launch more shards based on what error it returns.